File: examples/validation/validate-dossier.py
#!/usr/bin/env python3

"""
Dossier Schema Validator (Python)

Validates Dossier frontmatter against the JSON Schema.

Usage:
    pip install jsonschema
    python validate-dossier.py /path/to/dossier.md
"""

import sys
import json
import re
from pathlib import Path
from jsonschema import validate, ValidationError, Draft7Validator


def extract_frontmatter(file_path):
    """Extract JSON frontmatter from a Dossier markdown file."""
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Match ---dossier\n{...}\n---
    pattern = r'^---dossier\s*\n([\s\S]*?)\n---'
    match = re.search(pattern, content, re.MULTILINE)

    if not match:
        raise ValueError(
            'No dossier frontmatter found. Expected:\n---dossier\n{...}\n---'
        )

    json_string = match.group(1)

    try:
        return json.loads(json_string)
    except json.JSONDecodeError as e:
        raise ValueError(f'Failed to parse frontmatter JSON: {e}')


def load_schema():
    """Load the Dossier JSON Schema."""
    script_dir = Path(__file__).parent
    schema_path = script_dir / '../../dossier-schema.json'

    with open(schema_path, 'r', encoding='utf-8') as f:
        return json.load(f)


def validate_dossier(file_path):
    """Validate a Dossier file."""
    print(f'\n🔍 Validating: {file_path}\n')

    try:
        # Extract frontmatter
        frontmatter = extract_frontmatter(file_path)
        print('✓ Frontmatter extracted successfully')
        print(f"  Title: {frontmatter.get('title', 'N/A')}")
        print(f"  Version: {frontmatter.get('version', 'N/A')}")
        print(f"  Status: {frontmatter.get('status', 'N/A')}")
        print()

        # Load schema
        schema = load_schema()

        # Validate against schema
        validator = Draft7Validator(schema)
        errors = list(validator.iter_errors(frontmatter))

        if not errors:
            print('✅ VALID - Dossier schema is compliant\n')
            return True
        else:
            print('❌ INVALID - Schema validation failed:\n')
            for i, err in enumerate(errors, 1):
                path = '.'.join(str(p) for p in err.path) or '(root)'
                print(f'  Error {i}:')
                print(f'    Path: {path}')
                print(f'    Message: {err.message}')
                if err.validator_value:
                    print(f'    Constraint: {err.validator} = {err.validator_value}')
                print()
            return False

    except Exception as e:
        print(f'❌ ERROR: {e}\n')
        return False


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        print('Usage: python validate-dossier.py <dossier-file.md>')
        print('Example: python validate-dossier.py ../../examples/devops/deploy-to-aws.md')
        sys.exit(1)

    file_path = Path(sys.argv[1])

    if not file_path.exists():
        print(f'Error: File not found: {file_path}')
        sys.exit(1)

    valid = validate_dossier(file_path)
    sys.exit(0 if valid else 1)


if __name__ == '__main__':
    main()

File: examples/validation/validate-dossier.js
#!/usr/bin/env node

/**
 * Dossier Schema Validator (Node.js)
 *
 * Validates Dossier frontmatter against the JSON Schema.
 *
 * Usage:
 *   npm install ajv ajv-formats
 *   node validate-dossier.js /path/to/dossier.md
 */

const fs = require('fs');
const path = require('path');
const Ajv = require('ajv');
const addFormats = require('ajv-formats');

// Load schema
const schemaPath = path.join(__dirname, '../../dossier-schema.json');
const schema = JSON.parse(fs.readFileSync(schemaPath, 'utf8'));

// Initialize Ajv validator
const ajv = new Ajv({ allErrors: true, verbose: true });
addFormats(ajv);
const validate = ajv.compile(schema);

/**
 * Extract JSON frontmatter from a Dossier markdown file
 */
function extractFrontmatter(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');

  // Match ---dossier\n{...}\n---
  const frontmatterRegex = /^---dossier\s*\n([\s\S]*?)\n---/m;
  const match = content.match(frontmatterRegex);

  if (!match) {
    throw new Error('No dossier frontmatter found. Expected:\n---dossier\n{...}\n---');
  }

  const jsonString = match[1];

  try {
    return JSON.parse(jsonString);
  } catch (err) {
    throw new Error(`Failed to parse frontmatter JSON: ${err.message}`);
  }
}

/**
 * Validate a Dossier file
 */
function validateDossier(filePath) {
  console.log(`\n🔍 Validating: ${filePath}\n`);

  try {
    // Extract frontmatter
    const frontmatter = extractFrontmatter(filePath);
    console.log('✓ Frontmatter extracted successfully');
    console.log(`  Title: ${frontmatter.title}`);
    console.log(`  Version: ${frontmatter.version}`);
    console.log(`  Status: ${frontmatter.status}`);
    console.log();

    // Validate against schema
    const valid = validate(frontmatter);

    if (valid) {
      console.log('✅ VALID - Dossier schema is compliant\n');
      return true;
    } else {
      console.log('❌ INVALID - Schema validation failed:\n');
      validate.errors.forEach((err, i) => {
        console.log(`  Error ${i + 1}:`);
        console.log(`    Path: ${err.instancePath || '(root)'}`);
        console.log(`    Message: ${err.message}`);
        if (err.params) {
          console.log(`    Details: ${JSON.stringify(err.params)}`);
        }
        console.log();
      });
      return false;
    }
  } catch (err) {
    console.log(`❌ ERROR: ${err.message}\n`);
    return false;
  }
}

/**
 * Main
 */
function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log('Usage: node validate-dossier.js <dossier-file.md>');
    console.log('Example: node validate-dossier.js ../../examples/devops/deploy-to-aws.md');
    process.exit(1);
  }

  const filePath = args[0];

  if (!fs.existsSync(filePath)) {
    console.error(`Error: File not found: ${filePath}`);
    process.exit(1);
  }

  const valid = validateDossier(filePath);
  process.exit(valid ? 0 : 1);
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { extractFrontmatter, validateDossier };
